### DISCUSSION TITLE
Data set in `mount` not available in subsequent calls

### DISCUSSION THREAD

**Discussion by ju5t** [livewire/livewire#3702] [1. Help]
I must be doing something wrong here, but I have a lot of trouble debugging Livewire. We have a component that can be used to create and edit user records.

The component looks like this:

```php
    public function mount(int $customerId, ?int $userId = null): void
    {
        $customer = Customer::find($customerId);
        Gate::authorize('update', $customer);

        $this->customer = $customer;

        $this->user = is_null($userId)
            ? User::make(['customer_id' => $customer->id])
            : User::query()->where(['id' => $userId, 'customer_id' => $customer->id])->first();
    }

    public function save(): void
    {
        $this->validate();

        if (! empty($this->password)) {
            $this->user->password = Hash::make($this->password);
        }

        (new SaveUserAction($this->user, false))->execute();

        $this->emit('refreshUsers');
        $this->closeModal();
    }
```

It has `$rules` set and we've type hinted both `$user` and `$customer` with its respective models.

The `SaveUserAction` is a wrapper for some business logic, at it's core it's just doing a `$user->save()`. We do this all over the place, never had any issues with it. It's 100% not the reason for the errors we're seeing either.

I'm testing this with:

```php
        Livewire::actingAs($user)
            ->test(UserEditModal::class, ['customerId' => $customer->id])
            ->set('user.name', '::name::')
            ->set('user.email', 'name@domain.com')
            ->set('password', '::password::')
            ->set('password_confirmation', '::password::')
            ->call('save')
            ->assertHasNoErrors()
            ->assertEmitted('refreshUsers');

        $this->assertDatabaseHas(User::class, [
            'name' => '::name::',
            'email' => 'name@domain.com',
            'customer_id' => $customer->id,
        ]);
```

The final assertion fails; it only found a record that looked similar, but doesn't have its customer id set. I can see this is not an error with the business logic in `SaveUserAction` as in `save`, before running the action, the user seems to have no `customer_id` anymore. When debugging `mount` the `customer_id` is in the user object.

`customer_id` is in the fillable array of the `User`-model too.

The obvious solution here would be to add the `customer_id` back when saving the customer. But I'm really surprised why this data isn't passed on as expected. What am I doing wrong?

**Comment by jonneroelofs** [comment]
I think Livewire discards the customer_id on User when dehydrating because it it is not in your rules. 
I have ran into this problem as well. I have stopped setting the foreign key in the mount method and do just as you have suggested yourself: set the foreign key just before saving. 

This is the source code that dehydrates models:
![image](https://user-images.githubusercontent.com/20800818/132949170-7794449b-ff61-4969-add5-be8a0ff60ac2.png)

**Comment by ju5t** [comment]
@jonneroelofs thanks for the suggestion, changing the rules makes no difference unfortunately.

I am not that into the internals of Livewire but I assumed that having the data set in `mount()` would make it persistent. 

What's weird is this:

- Our User model has a single default `$attributes` value set. It's `disabled` and set to false by default.
- In `mount()`, when I set `$this->user` to `User::make([...])`, both `disabled` and `customer_id` are set.
- In `save()`, `disabled` is still set, but `customer_id` is gone.
- Neither are in the `rules` array.

🤯 

Setting `customer_id` in `save()` is ok, but I don't understand why the initial idea doesn't work.

**Comment by joshhanley** [comment]
@ju5t the reason that disabled is set, is because that will be the default coming from your model. 

So Livewire will only track properties that have rules for them and send that data to the front end, everything else comes from whatever is set when `User::make()` is called without any properties passed in, hence why your disabled property works.

As suggested I would be populating the `customer_id` in your save method, as you don't want the frontend to be able to modify this property.

Hope this helps! Let me know if you want me to explain anything else 🙂

**Reply by ju5t** [reply]
@joshhanley got it! I must have done something wrong when I added it to the rules, as that didn't seem to work.

Either way, having it in the `save` method is safer!
